lib/p2p/packets/Packet.ts

```diff
@@ -1,11 +1,36 @@
 import PacketType from './PacketType';
+import CryptoJS from 'crypto-js';
+import SHA256 from 'crypto-js/sha256';
+
+type PacketHeader = {
+  hash?: string;
+  reqHash?: string;
+};
+
+enum MessageType {
+  UNILATERAL,
+  REQUEST,
+  RESPONSE,
+}
 
 abstract class Packet {
+  public header: PacketHeader;
   public abstract type: PacketType;
   public abstract body: any;
-  public abstract responseType?: PacketType;
-  public abstract responseTimeout?: number;
-  public abstract toRaw(): string;
+  public abstract messageType: MessageType;
+
+  constructor (body: any, header?: PacketHeader) {
+    this.header = header || {
+      hash: SHA256(body).toString(CryptoJS.enc.Hex),
+      ...header!,
```

Comment:
What is the purpose of the `...header!` line if header is null?

Reply:
Good catch - it was a bug, after some refactoring which I did. 

I changed it to: `hash: (header && header.hash) || SHA256...`
The reason for this clumsy statement is to support these functionalities:
* it's an incoming packet deserialization, so the full header is provided
* it's a new packet being constructed. partial header might be given (`reqHash` field), or not

I can try to split the two modes somehow, if the code is too confusing.